### DISCUSSION TITLE
Tasmota rules - can't fire up trigger while sensor changes

### DISCUSSION THREAD

**Discussion by zmechu** [arendst/Tasmota#14674] [Q&A] [answered]
Hello!

Seems to be a simple, but I can't handle it :-)

Got tasmota 10.1.0(sensors) on Wemos D1 Mini. Works fine. GPIO17 is ADC Input - tasmota reads ADC and shows value on web interface, telemetry. But ADC value is sent to MQTT onli in "telemetry period", but I want to sent on every changes.

So, when ADC is a constant value, should be sent with telemetry.
But when ADC changes, it should be sent immediately.

Telemetry sent:
`tele/tasmota_1F536B/SENSOR = {"Time":"2022-01-30T11:43:33","ANALOG":{"A0":7}}`

so, prepare a Rule1:
`Rule1 ON ANALOG#A0 DO Publish tele/tasmota_1F536B/SENSOR {"Time":"%timestamp%","ANALOG":{"A0":%value%}} ENDON`
which seems to be almost fine - rule firing, but every ~4 seconds, when ADC doesn't changes.

Trying “A0#State” and “ANALOG#State” didn't help. "ANALOG#A0div10" sends me wrong value (value div 10).

What I'm doing wrong?

**Comment by barbudor** [comment, marked as answer]
Not searching enough in the provided examples
https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached

**Comment by zmechu** [comment]
I know it, but:
1. It returns me not sensor value, but value "div10"
2. I don't want to detect any "delta" value, just detect any value change
3. div10 didn't detect sensor value change from 0 to 10 (div10=0) or from 87 to 97 (div10=7) 
4. want to keep rules "as simple as possible". Just want to detect sensor change, without any delta, hysteresis, on single rule - and so on.

**Reply by barbudor** [reply]
```
ON ANALOG#A0!=var1 DO backlog var1 %value%; Publish tele/tasmota_1F536B/SENSOR {"Time":"%timestamp%","ANALOG":{"A0":%value%}} ENDON
```
Just the same as in the given example but simplified without hysteresys.

Note that due to the noise of the ESP ADC, it's likely that this will be almost equivalent to publish every value without testing.
I've never seen the ADC giving twice the save value.

**Reply by zmechu** [reply]
Still the same... Seems ADC problems, but it is strange. 10bit ADC with resistor connected to ADC input shouldn't result with so big noise...
```
13:51:51.861 RUL: ANALOG#A0!=VAR1 performs "backlog var1 744; Publish tele/tasmota_1F536B/SENSOR {"Time":"2022-01-30T13:51:51","ANALOG":{"A0":744}}"
13:51:51.872 MQT: stat/tasmota_1F536B/RESULT = {"Var1":"744"}
13:51:52.071 MQT: tele/tasmota_1F536B/SENSOR = {"Time":"2022-01-30T13:51:51","ANALOG":{"A0":744}}
13:51:56.207 RUL: ANALOG#A0!=VAR1 performs "backlog var1 744; Publish tele/tasmota_1F536B/SENSOR {"Time":"2022-01-30T13:51:56","ANALOG":{"A0":744}}"
13:51:56.218 MQT: stat/tasmota_1F536B/RESULT = {"Var1":"744"}
13:51:56.468 MQT: tele/tasmota_1F536B/SENSOR = {"Time":"2022-01-30T13:51:56","ANALOG":{"A0":744}}
13:52:00.449 RUL: ANALOG#A0!=VAR1 performs "backlog var1 744; Publish tele/tasmota_1F536B/SENSOR {"Time":"2022-01-30T13:52:00","ANALOG":{"A0":744}}"
13:52:00.462 MQT: stat/tasmota_1F536B/RESULT = {"Var1":"744"}
13:52:00.710 MQT: tele/tasmota_1F536B/SENSOR = {"Time":"2022-01-30T13:52:00","ANALOG":{"A0":744}}
13:52:04.638 RUL: ANALOG#A0!=VAR1 performs "backlog var1 744; Publish tele/tasmota_1F536B/SENSOR {"Time":"2022-01-30T13:52:04","ANALOG":{"A0":744}}"
13:52:04.720 MQT: stat/tasmota_1F536B/RESULT = {"Var1":"744"}
13:52:04.938 MQT: tele/tasmota_1F536B/SENSOR = {"Time":"2022-01-30T13:52:04","ANALOG":{"A0":744}}
```

**Reply by barbudor** [reply]
My bad you need '%' around var1 in the comparison

**Reply by zmechu** [reply]
Rule
`Rule1 ON ANALOG#A0!=%var1% DO backlog var1 %value%; Publish tele/tasmota_1F536B/SENSOR {"Time":"%timestamp%","ANALOG":{"A0":%value%}} ENDON`
Seems to works fine :-)
Two questions:
1. Why simple "ANALOG#A0" doesn't works?
2. As I see ADC "query" time is about 4 seconds - when ADC changes, tasmota show changes not immediately, but after few seconds. Is there a way to change ADC speed?

**Reply by barbudor** [reply]
The ADC is measured every 250ms
But there is an hysteresis so not change below 10 (ESP8266) or 41 (ESP32) will be forwarded to the Rules engine

https://github.com/arendst/Tasmota/blob/405d1e6eb97929659ab60aae9f60249e695d5567/tasmota/xsns_02_analog.ino#L334-L342

You can't change from Settings. If this is the behavior you want, you will have to change the code and build your own version.

**Reply by zmechu** [reply]
OK, till nowe seems to be working! Thank you for help!

**Reply by Patryk051185** [reply]
Will this work for esp32. I tried to apply it and do a build and unfortunately it doesn't work. I use Esp32 with a pressure transducer for the water tank and I get a reading every 10 seconds, not as I would like immediately after the status change. I don't want to damage the pump by running it dry.

**Reply by barbudor** [reply]
Not knowing what exact change you've done or what is your rule, hard to say what goes wrong
As written above, the measurement is taken every 250ms so the proper rule should catch the 'empty' state within that time.
Getting a reading every 10 seconds let me think that you are using the 'tele-' prefix in your rule with a teleperiod reduced to 10 seconds

**Reply by Patryk051185** [reply]
I did exactly the same as above according to the posts.
Tele Period set to 10 seconds because that's the least amount of time.
With a fast-running pump, the condition can change quickly in less than 10 seconds.
I compiled the build with the suggestions from the post https://github.com/arendst/Tasmota/discussions/14674#discussioncomment-2076085

**Reply by sfromis** [reply]
Did you make sure to also enable the rule set, instead of only having it saved. It takes a 2nd command, like `Rule1 1`

In general, better show what you actually did, **and** what the logs says, instead of only suggesting that you did "the same".

**Reply by barbudor** [reply]
> I compiled the build with the suggestions from the post https://github.com/arendst/Tasmota/discussions/14674#discussioncomment-2076085

That's my message and there is no code change listed. Only a suggetsion on the location where such change may be done.

**Reply by Patryk051185** [reply]
I'm a newbie at this. I thought that no changes were made, comparing them with what was in the original build and in post #14674 (reply in thread) because when I looked at them they looked the same. Can I ask you for a hint on what to change in this code? I would just like to achieve the effect of refreshing the state immediately after the state changes, and not every 10 seconds.
